packages/components/src/DateTimePickers/DateRange/Manager/Manager.component.js

```diff
@@ -0,0 +1,188 @@
+import React, { useState, useEffect } from 'react';
+import PropTypes from 'prop-types';
+import isAfter from 'date-fns/is_after';
+import isBefore from 'date-fns/is_before';
+
+import { DateRangeContext } from '../Context';
+import { START_DATE, END_DATE } from '../constants';
+import {
+	extractDate,
+	extractFromDate,
+	extractPartsFromTextInput,
+} from '../../Date/date-extraction';
+
+function extractRangeParts(startDate, endDate, options) {
+	const startDateParts = extractDate(startDate, options);
+	const endDateParts = extractDate(endDate, options);
+
+	return {
+		startDate: {
+			value: startDateParts.localDate,
+			textInput: startDateParts.textInput,
+		},
+		endDate: {
+			value: endDateParts.localDate,
+			textInput: endDateParts.textInput,
+		},
+		errors: startDateParts.errors.concat(endDateParts.errors),
+		errorMessage: startDateParts.errorMessage || endDateParts.errorMessage,
+	};
+}
+
+function ContextualManager(props) {
+	const options = {
+		dateFormat: props.dateFormat,
+	};
+	const initialState = extractRangeParts(props.startDate, props.endDate, options);
+	const [state, setState] = useState(initialState);
+
+	useEffect(() => {
+		if (props.startDate !== state.startDate.value || props.endDate !== state.endDate.value) {
+			const parts = extractRangeParts(props.startDate, props.endDate, options);
+			setState(parts);
+		}
+	}, [props.startDate, props.endDate]);
+
+	function onDatesChange(event, nextState) {
+		if (props.onChange) {
+			const payload = {
+				startDate: nextState.startDate.value,
+				endDate: nextState.endDate.value,
+				errors: nextState.errors,
+				errorMessage: nextState.errorMessage,
+				origin: 'RANGE_PICKER',
+			};
+			props.onChange(event, payload);
+		}
+	}
+
+	function onFocusChange(event, focusedInput) {
+		setState(prevState => ({
+			...prevState,
+			focusedInput,
+		}));
+	}
+
+	function onStartChange(event, { date: startDate }) {
+		const { date, textInput } = extractFromDate(startDate, options);
+		const nextState = {};
+
+		// clear endDate when new startDate is after existing endDate
+		if (state.endDate.value && isAfter(date, state.endDate.value)) {
+			nextState.endDate = {
+				value: undefined,
+				textInput: '',
```

Comment:
Never change the user input.
Let the value and textInput from extract, but add an error. Let's try to clean that
- onStartChange: extract new start date + call common onChange with origin = `START_PICKER` (if it is picked)
- onEndChange: extract new end date + call common onChange with origin = `END_PICKER` (if it is picked)
- common onChange: check that the dates are consistent (start before end) and manage the errors.

Reply:
changed as suggested.
move the check of start/end date inside `onChange`.